django/db/models/deletion.py

```diff
@@ -265,6 +265,7 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
         if keep_parents:
             parents = set(model._meta.get_parent_list())
         model_fast_deletes = defaultdict(list)
+        protected_objects = {'keys': [], 'sub_objs': []}
```

Comment:
I don't see much value in this structure, why not use a dictionary?
```python
protected_objects = defaultdict(list)
...
            try:
                field.remote_field.on_delete(self, field, sub_objs, self.using)
            except ProtectedError:
                key = "'%s.%s'" % (sub_objs[0].__class__.__name__, field.name)
                protected_objects[key] = sub_objs
...
if protected_objects:
    raise ProtectedError(
        'Cannot delete some instances of model %r because they are '
        'referenced through protected foreign keys: %s.' % (
            field.remote_field.model.__name__,
            ', '.join(protected_objects),
        ),
        protected_objects.values(),
    )
```

Reply:
@felixxm As far as I know, returned objects in `ProtectedError` should be a list of objects(sub_objs). That's why I used that structure.
I replaced your suggested structure and you can see the errors.

Reply:
I've added snippet with an idea not a tested solution. I've [pushed](https://github.com/django/django/compare/4df976975818fa610c1aa397b3bc3f03413790f6..4f100dcdb6fb43f11cf560997287ddac7cb10380) fix.

Reply:
Sorry for my misunderstanding.
Thanks for the fix.